design/proposed/2020-openstackserver-controller.md

```diff
@@ -0,0 +1,160 @@
+
+Creating a Machine/OpenStackMachine object for the bastion could negatively impact other CAPI users.
+
+## Create a new controller for the bastion
+
+This is essentially the same as using OpenStackMachine, except using a new CRD which can be independent of other CAPI objects.
+
+The obvious disadvantage is that it requires a new CRD and controller.
+
+# Proposal: An OpenStackServer CRD and controller
+
+## Philosophy
+
+We will create a new controller which is capable, in principal, of doing all of the OpenStack resource creation tasks common to bastion and OpenStackMachine creation. That is:
+
+* Resolve OpenStack resource parameters to specific resources (e.g. image, networks, subnets, security groups, server groups)
+* Create and delete ports
+* Create and delete volumes
+* Create and delete the server
+
+It will have no dependency on Machine or OpenStackCluster. So things it will explicitly not do include:
+
+* Attaching the Bastion floating IP
+* Creating a loadbalancer member
+* Adding a port for the default cluster network if no ports are provided
+* Referencing an AvailabilityZone from the Machine object
+* Adding default tags from the OpenStackCluster object
+
+## API
+
+Based on OpenStackMachineSpec, with modifications to accomodate the above.
+
+
+```go
+type OpenStackServerSpec struct {
+	AvailabilityZone       string
+	Flavor                 string
+	Image                  ImageParam
+	SSHKeyName             string
+	Ports                  []PortOpts
+	SecurityGroups         []SecurityGroupParam
+	Trunk                  bool
+	Tags                   []string
+	ServerMetadata         []ServerMetadata
+	ConfigDrive            *bool
+	RootVolume             *RootVolume
+	AdditionalBlockDevices []AdditionalBlockDevice
+	ServerGroup            *ServerGroupParam
+	IdentityRef            *OpenStackIdentityReference
+	FloatingIPPoolRef      *corev1.TypedLocalObjectReference
+	UserDataRef            *corev1.TypedLocalObjectReference
+}
+
+type OpenStackServerStatus struct {
+	InstanceID    optional.String
+	InstanceState *InstanceState
+	Addresses     []corev1.NodeAddress
+	Resolved      *ResolvedMachineSpec
+	Resources     *MachineResources
+	Conditions    clusterv1.Conditions
+}
+```
+
+As the new API is non-trivial we should initially create it in v1alpha1.
+
+### Upgrading an existing deployment
+
+OpenStackServer would be a new API and does not affect any existing API.
+
+In the same way that OpenStackMachine currently has an 'adoption' phase where it will adopt existing resources, OpenStackServer should adopt matching resources which it would have created. On upgrade to a new version of CAPO which manages the bastion with an OpenStackServer object, I would expect the flow to be:
+
+* Cluster controller creates OpenStackServer and waits for it to report Ready
+* OpenStackServer controller looks for existing resources matching its given spec
```

Comment:
This part is a bit tricky, we'll need to compare every field out of instanceStatus, I started to play with it and it became messy.
Not against this idea, just making sure we really want to go down that path.

The risk of not doing that is that an instance would be adopted by OpenStackServer that wouldn't be matching 100% the spec if the instance would have been created by the OpenStackServer controller.

Reply:
I was thinking the adoption behaviour would be *identical* to what we currently do in the OpenStackMachine controller. Literally the same code. We're basically just adopting by name.